cranelift/src/disasm.rs

```diff
@@ -69,75 +69,8 @@ pub fn print_stack_maps(traps: &[MachStackMap]) -> String {
-                    .mode(arch::x86::ArchMode::Mode32)
-                    .build()
-                    .map_err(map_caperr)?,
-                Architecture::X86_64 => Capstone::new()
-                    .x86()
-                    .mode(arch::x86::ArchMode::Mode64)
-                    .build()
-                    .map_err(map_caperr)?,
-                Architecture::Arm(arm) => {
-                    if arm.is_thumb() {
-                        Capstone::new()
-                            .arm()
-                            .mode(arch::arm::ArchMode::Thumb)
-                            .build()
-                            .map_err(map_caperr)?
-                    } else {
-                        Capstone::new()
-                            .arm()
-                            .mode(arch::arm::ArchMode::Arm)
-                            .build()
-                            .map_err(map_caperr)?
-                    }
-                }
-                Architecture::Aarch64 {..} => {
-                    let mut cs = Capstone::new()
-                        .arm64()
-                        .mode(arch::arm64::ArchMode::Arm)
-                        .build()
-                        .map_err(map_caperr)?;
-                    // AArch64 uses inline constants rather than a separate constant pool right now.
-                    // Without this option, Capstone will stop disassembling as soon as it sees
-                    // an inline constant that is not also a valid instruction. With this option,
-                    // Capstone will print a `.byte` directive with the bytes of the inline constant
-                    // and continue to the next instruction.
-                    cs.set_skipdata(true).map_err(map_caperr)?;
-                    cs
-                }
-                Architecture::S390x {..} => Capstone::new()
-                    .sysz()
-                    .mode(arch::sysz::ArchMode::Default)
-                    .build()
-                    .map_err(map_caperr)?,
-                Architecture::Riscv64 {..} => {
-                    let mut cs = Capstone::new()
-                        .riscv()
-                        .mode(arch::riscv::ArchMode::RiscV64)
-                        .build()
-                        .map_err(map_caperr)?;
-                    // Similar to AArch64, RISC-V uses inline constants rather than a separate
-                    // constant pool. We want to skip dissasembly over inline constants instead
-                    // of stopping on invalid bytes.
-                    cs.set_skipdata(true).map_err(map_caperr)?;
-                    cs
-                }
-                _ => anyhow::bail!("Unknown ISA"),
-            };
-
-            Ok(cs)
-        }
-
         pub fn print_disassembly(isa: &dyn TargetIsa, mem: &[u8]) -> Result<()> {
-            let cs = get_disassembler(isa)?;
+            let cs = isa.to_capstone().map_err(|e| anyhow::format_err!("{}", e))?;
```

Comment:
Potential future improvement below (I don't want to necessarily block the PR on this!): the below disassembly logic is largely similar to `MachBuffer`'s now, except that it prints bytes too, and doesn't include relocs, and currently has just a slice-of-`u8` here; we could make those format differences options, wire through a `MachBuffer` somehow, and delegate to the common code here.

Reply:
I like that, and would be happy to push that change through before merging this :+1:

Reply:
I looked into this, and it would be a bit more work than I'd like to tackle in this PR. I'll make a follow-up pr :+1: